File: com/generic/Timer/Timer.java
package com.generic.Timer;

public class Timer extends Thread {
    private boolean stop;
    private boolean chrono;
    // ? true = chrono | false = décompte

    private int tick;
    // ? Compte toutes les [interval]. (ms)
    private int tMax;
    // ? Valeur d'origine du décompte (-1 si chrono) (ms)
    private int currT;
    // ? temps actuellement écoulé. (ms)

    private TimerLink tl;

    public Timer(boolean chrono, int tick, int tMax, TimerLink tl) {
        this.chrono = chrono;
        this.tick = tick;
        this.tMax = tMax;
        this.tl = tl;
    }

    public static Timer createChrono(int tick) {
        return new Timer(true, tick, -1, null);
    }

    public static Timer createCounter(int val, int tick, TimerLink tl) {
        return new Timer(false, tick, val, tl);
    }

    public void run() {
        try {
            while (!stop) {
                currT += tick;
                if (currT >= tMax && !chrono) {
                    stop = false;
                    tl.timerEnded();
                }

                sleep(tick);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void stopTimer() {
        stop = true;
    }

    public int getCurrentTime() {
        return this.currT;
    }
}

File: csharp_common/csharp_common/Timer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

namespace csharp_common
{
    public class Timer
    {
        private Thread counterThread;
        private bool stop{ get; set; }

        private bool chrono { get; set; }

        private int tick { get; set; }

        private int tMax { get; set; }

        private int currT;

        private ITimer toNotify { get; set; }

        private Timer(bool chrono, int tick, int tMax, ITimer it)
        {
            this.chrono = chrono;
            this.tick = tick;
            this.tMax = tMax;
            this.toNotify = it;

            this.counterThread = new Thread(new ThreadStart(this.run));
        }

        public static Timer createChrono(int tick)
        {
            return new Timer(true, tick, -1, null);
        }

        public static Timer createCounter(int maxT, int tick, ITimer it)
        {
            return new Timer(false, tick, maxT, it);
        }

        public void start()
        {
            counterThread.Start();
        }

        public void run()
        {
            while(!stop)
            {
                currT += tick;
                if (currT >= tMax && chrono)
                {
                    stop = false;
                    toNotify.timerEnded();
                }

                Thread.Sleep(tick);
            }
        }

        public void stopTimer()
        {
            stop = true;
        }

        public int getTime()
        {
            return this.currT;
        }
    }

    public interface ITimer
    {
        void timerEnded();
    }
}
